### PR TITLE
Add ETOR support to RESTTransport

### PR DESCRIPTION
**Pull request by oslynn**
This PR adds ETOR support to RESTTransport for Jimmy's latest code.  The transport allows the sender/Lab/Hospital to report Test Result to ETOR/Trusted-Intermediary.

Test Steps:
1. ./gradlew clean package
2. ./prime test

## Changes
- *Include a comprehensive list of changes in this PR*
- *(For web UI changes) Include screenshots/video of changes*

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- Fixes #issue
- https://app.zenhub.com/workspaces/engagement-6166edbd409257001e09f1b2/issues/gh/cdcgov/prime-reportstream/9547

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (jimmyfagan) [inline prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:392]: This test fails for me, not sure if I'm missing some setup or something, but I think generally integration tests should work without any setup.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:1]: A lot of the logic in this file is depending on class-types instead of variables, and using hard coded logic for parameters and similar things. This is functionally okay, but I'm wondering if there's a better, more maintainable way of organizing this. For instance, maybe instead of having `UserApiKeyCredential`/`UserEtorCredential` and `getAuthTokenWithUserApiKey`/`getAuthTokenWithUserEtor`, could we just make the existing classes/functions  more flexible? Maybe `UserApiKeyCredential` has something like `var paramMap: Map<String, String>` and we build the form parameters from that map in `getAuthTokenWithUserApiKey`? Then we wouldn't need any of the etor specific classes or functions, we could just populate that map as part of the deserialization.

**Review by jimmyfagan — COMMENTED** [review]
I have some comments, questions, and suggestions. You should also fill out the description in the PR to give more context to reviewers. Some basic test steps, a description of the changes you made, and some light context around why you made some of the changes you did would be very helpful.

File: prime-router/src/main/kotlin/credentials/Credential.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/credentials/Credential.kt:82]: Yes, we can.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:447]: Yes, I totally agreed with you.  I will create a ticket to do so.

File: prime-router/settings/staging/0153-add-etor-service-receiver.yml
Comment (oslynn) [reply prime-router/settings/staging/0153-add-etor-service-receiver.yml:1]: Ok, if that is what you want.  I will edit the existing 0149-etor.yml.  In the past when we added a new setting, other people showed me to add a new updated file and keep the old one as it is.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:385]: It is fixed, now.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:1]: Yes, I will remove the UserEtorCredential class and only UserApiKeyCredential.  Adding paramMap though, I will have to update all the existing credentials in the Vault.  I will look into this as well.

File: prime-router/settings/staging/0153-add-etor-service-receiver.yml
Comment (jimmyfagan) [reply prime-router/settings/staging/0153-add-etor-service-receiver.yml:1]: Ah I see, yea I could see that being the right approach in some situations, but I think in our case, I'd appreciate having all the ETOR settings in that file, I've been keeping that updated with things like the `etor-ti` topic already. I could see having separate receivers for REST and for SFTP in 0149-etor.yml for now if we want to preserve the SFTP until the REST transport is more stable, but keeping it in the same file makes it easier to find and easier to keep track of what the most up-to-date settings are.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (jimmyfagan) [inline prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:392]: What do these values correspond to? Or are they just random?

**Review by jimmyfagan — COMMENTED** [review]
Thanks for meeting with me to talk about this more! Left some comments based on our conversation and also a few more questions. Thanks for the adjustments so far, looking pretty good!

**Comment by oslynn** [comment]
I have new PR for this.

File: prime-router/src/main/kotlin/TransportType.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/TransportType.kt:120]: If the intention is that this is never null, the `?` should be removed so that'll be enforced by the compiler.
```suggestion
    val parameters: Map<String, String> = emptyMap(),
```

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:262]: Did you run `./gradlew ktlintFormat`? Do you have your precommit checks set up? I wouldn't have been able to check this in because it doesn't conform to our code format restrictions, I _think_ those checks can be installed with `.environment/githooks.sh install`.
```suggestion
                bearerTokens = BearerTokens(tokenInfo.accessToken, tokenInfo.refreshToken ?: "")
```

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:339]: If you change `parameters` to be non-nullable, then you can remove the `!!`
```suggestion
                    if (restTransportInfo.parameters.isEmpty()) {
```
If you don't make `parameters` non-nullable, then maybe something like `restTransportInfo.parameters?.isEmpty() != false` would be safer.

File: prime-router/src/main/kotlin/transport/TokenInfo.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/transport/TokenInfo.kt:1]: This looks good, but you can delete `TokenInfoEtor` now!

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (jimmyfagan) [inline prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:22]: This is another thing that our pre-commit checks don't allow, be sure to run `./gradlew ktlintFormat` before you commit!

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (jimmyfagan) [inline prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:387]: We can use the default null values for the middle 3 parameters
```suggestion
                TokenInfo(accessToken = "MockToken", tokenType = "bearer")
```

**Review by jimmyfagan — COMMENTED** [review]
Thanks for making the changes, things look pretty good! Left a few comments for some minor suggestions and formatting issues, but I think we're almost ready to merge.

File: prime-router/src/main/kotlin/TransportType.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/TransportType.kt:120]: Old REST Transports (i.e NY, OK, WA, .. etc) have null as Parameters.  I intentionally have it for backward compatibility with the old RESTransport.  Currently, only REST Transport for ETOR has parameters.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:262]: I will run the command, now.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:339]: As I explained above, parameters can be null.

File: prime-router/src/main/kotlin/transport/TokenInfo.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/TokenInfo.kt:1]: Ah, I must forget to remove, it.

File: prime-router/src/main/kotlin/TransportType.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/TransportType.kt:120]: Using the default value `emptyMap()` here means that even for those old transport, the Map should be an empty map, but still not `null`. We should test that to make sure, but given that there is a non-null default value, the only way for this to be `null` would be to have it explicitly set.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:339]: If `parameters` can be `null`, this check will fail due a null pointer exception. `!!` asserts that the variable is _not_ null, and will throw an exception if that is untrue. Based on the other conversation, since the default for `parameters` is `emptyMap()` I don't think this will actually be `null`, but if you think it might, then we should still use `restTransportInfo.parameters?.isEmpty() != false` to prevent execptions.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt]: Done.

File: prime-router/src/main/kotlin/transport/TokenInfo.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/TokenInfo.kt:1]: Removed.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt]: Done.  IntelliJ got me, again!

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt]: Done

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt]: Random.

File: prime-router/src/main/kotlin/TransportType.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/TransportType.kt:120]: I set this to default as emptyMap().  If transport doesn't provide the parameters, it will default as an emptyMap().  I am sorry in the earlier response I used null instead of emptyMap().  In the code, it checks for .isEmpty() or not.  I tested a transport with parameters and without parameters and it works.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:339]: Yes, you are correct.  As I mentioned above, I used null wrongly.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt]: Ok, I will change it.  I wonder why since they are defaulting to null already in the TokenInfo object?

File: prime-router/src/main/kotlin/TransportType.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/TransportType.kt:120]: I understand, since it is not intended to ever be null, the `?` should be removed so that will be enforced by the compiler.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:339]: Got it, in that case, we should make the variable non-nullable and remove the `!!` here.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (jimmyfagan) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt]: Functionally it doesn't change anything, but I find it a lot more readable and less error-prone to use the parameter names and omit variables that use the defaults. But it's not critical.

File: prime-router/src/main/kotlin/transport/RESTTransport.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:424]: Is there a ticket out there for this yet?

Comment (oslynn) [reply prime-router/src/main/kotlin/transport/RESTTransport.kt:424]: I just create ticket #10119 to move this hardcode out to the setting file instead.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt]: Done

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (victor-chaparro) [inline prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:391]: Is there a way we can test getAuthTokenWithUserApiKey without mocking it? RESTTransport.kt doesn't have any code coverage, because most of it is mocked. Maybe we can create an additional ticket to add that. But, I'm concerned that we don't have enough testing to make sure this doesn't break other receivers that use REST. Is there a way we can test this doesn't break that before merging?

**Review by victor-chaparro — COMMENTED** [review]
I spent some time trying to understand this code and it looks good to me, but I'm just wondering how we can test this locally and make sure it doesn't break existing functionality.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:391]: @victor-chaparro, thank you for reviewing the PR.  To verify the Transport working with ETOR/Trusted-Intermediary, I run the Trusted-Intermediary and ReportStream server locally with the setting provided in 0149-etor.yml (uncommented out the localhost).  Then, I manually send patient resource to the reportStream's localhost /api/report endpoint and check for a successful response.  However, this is not what I really want for testing the ETOR use case.  I agreed with you that the getAuthTokenWithUserApiKey needs to be tested without mock.  I will create a ticket to do so. 

For code coverage, I use IntelliJ run the test with the Coverage command and check for path coverage.

Comment (victor-chaparro) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:391]: Can we also test that it works for other receivers that use REST locally?

Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:391]: Yes, I can and will test with NY and OK to make certain that it works with them as well.

**Comment by oslynn** [comment]
Need to conduct a test with NY and OK to make sure it is working with them as well as ETOR.

**Review by victor-chaparro — APPROVED** [review]
I'm good with approving this as long as we've made sure it doesn't break other receivers, and tickets for improving code coverage have been created. Thanks Ott for all the work you've put into this.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:391]: Tested with NY and it work well.
image

**Review by jimmyfagan — APPROVED** [review]
Changes look good to me too, would just like to also confirm that we've tested this against Trusted Intermediary in staging to make sure we can send to them now that they enforce the auth tokens on the `demographics` endpoint.

**Comment by oslynn** [comment]
Tested with Trusted-Intermediary staging:

**Comment by jimmyfagan** [comment]
Thanks Ott, good work! I think we'll be good to close #9675 whenever this gets merged in!

**Comment by oslynn** [comment]
@jimmyfagan Before I merge, I would like to add two Unit Test Cases to cover "getAuthTokenWithUserApiKey" that I added for when Transport with parameters filled and without.

File: prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt
Comment (oslynn) [reply prime-router/src/testIntegration/kotlin/transport/RESTTransportIntegrationTests.kt:456]: Test cases to cover getAuthTokenWithUserApiKey are added to RESTTransportIntegrationTest.kt.

**Pull request merged**
